clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,135 @@
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+void UseStdMinMaxCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      ifStmt(
+          hasCondition(binaryOperator(hasAnyOperatorName("<", ">", "<=", ">="),
+                                      hasLHS(expr().bind("CondLhs")),
+                                      hasRHS(expr().bind("CondRhs")))),
+          hasThen(anyOf(stmt(binaryOperator(hasOperatorName("="),
+                                            hasLHS(expr().bind("AssignLhs")),
+                                            hasRHS(expr().bind("AssignRhs")))),
+                        compoundStmt(has(binaryOperator(
+                                         hasOperatorName("="),
+                                         hasLHS(expr().bind("AssignLhs")),
+                                         hasRHS(expr().bind("AssignRhs")))))
+                            .bind("compound"))))
+          .bind("if"),
+      this);
+}
+
+void UseStdMinMaxCheck::registerPPCallbacks(const SourceManager &SM,
+                                            Preprocessor *PP,
+                                            Preprocessor *ModuleExpanderPP) {
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void UseStdMinMaxCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *CondLhs = Result.Nodes.getNodeAs<Expr>("CondLhs");
+  const auto *CondRhs = Result.Nodes.getNodeAs<Expr>("CondRhs");
+  const auto *AssignLhs = Result.Nodes.getNodeAs<Expr>("AssignLhs");
+  const auto *AssignRhs = Result.Nodes.getNodeAs<Expr>("AssignRhs");
+  const auto *If = Result.Nodes.getNodeAs<IfStmt>("if");
+  const auto *Compound = Result.Nodes.getNodeAs<CompoundStmt>("compound");
+  const auto &Context = *Result.Context;
+  const auto &LO = Context.getLangOpts();
+  const SourceManager &Source = Context.getSourceManager();
+
+  const auto *BinaryOp = dyn_cast<BinaryOperator>(If->getCond());
+  if (!BinaryOp || If->hasElseStorage())
+    return;
+
+  if (Compound) {
+    if (Compound->size() > 1)
+      return;
+  }
+
+  const SourceLocation IfLocation = If->getIfLoc();
+  const SourceLocation ThenLocation = If->getEndLoc();
+
+  if (IfLocation.isMacroID() || ThenLocation.isMacroID())
+    return;
+
+  const auto CreateReplacement = [&](bool UseMax) {
+    const auto *FunctionName = UseMax ? "std::max" : "std::min";
+    const auto CondLhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(CondLhs->getSourceRange()), Source, LO);
+    const auto CondRhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(CondRhs->getSourceRange()), Source, LO);
+    const auto AssignLhsStr = Lexer::getSourceText(
+        Source.getExpansionRange(AssignLhs->getSourceRange()), Source, LO);
+    const auto AssignLhsType =
+        AssignLhs->getType().getCanonicalType().getAsString();
+    return (AssignLhsStr + " = " + FunctionName + "<" + AssignLhsType + ">(" +
```

Comment:
We should only specify the template type if the compiler is unable to deduce it. Otherwise it is redundant